adr/util/cache_stores.py

```diff
@@ -0,0 +1,97 @@
+import os
+import shutil
+import tarfile
+import tempfile
+from distutils.dir_util import copy_tree
+
+from cachy.stores import FileStore, NullStore  # noqa
+from loguru import logger
+
+from adr.util.req import requests_retry_session
+
+
+def extract_tar_zst(path, dest):
+    import zstandard
+    dctx = zstandard.ZstdDecompressor()
+    with open(path, "rb") as f:
+        with dctx.stream_reader(f) as reader:
+            with tarfile.open(mode="r|", fileobj=reader) as tar:
+                tar.extractall(dest)
+
+
+class SeededFileStore(FileStore):
+    RESEED_KEY = "adr:SeededFileStore:reseed"
+
+    def __init__(self, config):
+        """A FileStore instance that allows pre-seeding the cache from an archive
+        downloaded from a URL.
+
+        Example configuration:
+
+            file = {
+                driver = "seeded-file",
+                path = "/path/to/cache",
+                url = "https://example.com/cache.tar.gz"
+            }
+
+        Configuation can include:
+
+            path: Path on the local file system to store the cache.
+            url: Where to download the preseed data from.
+            archive_relpath: Path within the archive pointing to the root of the cache data.
+            reseed_interval: Time in minutes after which the data will be
+                              seeded again (defaults to no reseeding).
+
+        Supported archive formats include `.zip`, `.tar`, `.tar.gz`, `.tar.bz2` or `.tar.zst`.
+        """
+        self._url = config['url']
+        self._reseed_interval = config.get('reseed_interval', 0)
+        self._archive_relpath = config.get('archive_relpath')
+
+        self._session = requests_retry_session()
+
+        kwargs = {
+            'directory': config['path'],
+        }
+        if 'hash_type' in config:
+            kwargs['hash_type'] = config['hash_type']
+
+        super(SeededFileStore, self).__init__(**kwargs)
+
+    def seed(self):
+        """Download and extract the seed data to the cache directory."""
+        logger.info(f"Seeding adr cache at {self._directory} with contents from {self._url}")
+        self._create_cache_directory(self._directory)
+
+        filename = self._url.split('/')[-1]
+        with self._session.get(self._url, stream=True) as r:
+            r.raise_for_status()
+
+            try:
+                tempdir = tempfile.mkdtemp()
```

Comment:
Nit: you can use TemporaryDirectory to avoid the try-finally

Reply:
Done.